### PR TITLE
PT-13154 - Deactivate Sofort payment method

### PR DESCRIPTION
**Pull request by DennisGarding**

### CODE REVIEW

File: Components/PaymentMethodProvider.php [hunk 1]
```diff
@@ -187,6 +187,17 @@ public function getPaymentTypeByName($paymentMethodName)
         );
     }
 
+    /**
+     * @return array<self::*>
```
Comment (mitelg) [inline Components/PaymentMethodProvider.php:191]:
```suggestion
     * @return list<self::*>
```

File: Subscriber/Backend.php [hunk 2]
```diff
@@ -128,6 +128,7 @@ public function onPostDispatchPayment(ActionEventArgs $args)
         $view->addTemplateDir($this->pluginDir . '/Resources/views/');
 
         if ($args->get('request')->getActionName() === 'load') {
+            $view->assign('deactivatedPaymentMethods', PaymentMethodProvider::getDeactivatedPaymentMethods());
```
Comment (mitelg) [inline Subscriber/Backend.php:131]: would it be easier to explode the array directly to a comma separated string? :thinking: so you do not need to en- and decode to JSON in ExtJs

File: Setup/PaymentModels/PaymentInstaller.php [hunk 3]
```diff
@@ -42,7 +42,15 @@ public function __construct(
 
     public function installPayments()
     {
-        foreach (PaymentMethodProvider::getAllUnifiedNames() as $paymentMethodName) {
+        $deactivatedPaymentMethods = PaymentMethodProvider::getDeactivatedPaymentMethods();
+        $paymentMethods = \array_filter(
+            PaymentMethodProvider::getAllUnifiedNames(),
+            function ($paymentMethodName) use ($deactivatedPaymentMethods) {
+                return !\in_array($paymentMethodName, $deactivatedPaymentMethods);
```
Comment (mitelg) [inline Setup/PaymentModels/PaymentInstaller.php:49]: the third parameter should be provided to be type safe

File: Subscriber/ApmRiskManagement.php [hunk 4]
```diff
@@ -135,19 +134,26 @@ public function onExecuteApmRule(Enlight_Event_EventArgs $args)
         $values = $this->valueFactory->createValue($paymentType, $basket, $user);
         $validator = $this->validatorFactory->createValidator($paymentType);
 
-        // This is temporary to disable Trustly as it will not available until the end of quarter 2 2023.
-        if ($paymentType === PaymentType::APM_TRUSTLY) {
+        $deactivatedPaymentTypes = $this->getDeactivatedPaymentTypes();
+        if (\in_array($paymentType, $deactivatedPaymentTypes)) {
```
Comment (mitelg) [inline Subscriber/ApmRiskManagement.php:138]: also here, third parameter

File: Subscriber/ApmRiskManagement.php [hunk 5]
```diff
@@ -135,19 +134,26 @@ public function onExecuteApmRule(Enlight_Event_EventArgs $args)
...
+        $violationList = $this->validator->validate($values, $validator);
 
-            $violationList = $this->validator->validate($values, $validator, [$group]);
-        } else {
-            $violationList = $this->validator->validate($values, $validator);
+        return $violationList->count() > 0;
+    }
+
+    /**
+     * @return array<int, string>
```
Comment (mitelg) [inline Subscriber/ApmRiskManagement.php:148]:
```suggestion
     * @return list<PaymentType::*>
```

**Review by mitelg — CHANGES_REQUESTED** [review]

**Comment by mitelg** [comment]
just smaller things, nice work!

File: Subscriber/Backend.php (hunk 2 above)
Comment (DennisGarding) [reply Subscriber/Backend.php:131]: I dont know if this brings a benefit. Because you have to implode the string and explode the string to an array to.

**Review by mitelg — APPROVED** [review]

**Pull request merged**